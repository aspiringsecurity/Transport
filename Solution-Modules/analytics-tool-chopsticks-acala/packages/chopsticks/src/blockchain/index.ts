import { ApplyExtrinsicResult } from '@polkadot/types/interfaces'
import { DataSource } from 'typeorm'
import { HexString } from '@polkadot/util/types'
import { RegisteredTypes } from '@polkadot/types/types'
import { blake2AsHex } from '@polkadot/util-crypto'
import { u8aConcat, u8aToHex } from '@polkadot/util'
import type { TransactionValidity } from '@polkadot/types/interfaces/txqueue'

import { Api } from '../api'
import { Block } from './block'
import { BuildBlockMode, BuildBlockParams, DownwardMessage, HorizontalMessage, TxPool } from './txpool'
import { HeadState } from './head-state'
import { InherentProvider } from './inherent'
import { StorageValue } from './storage-layer'
import { compactHex } from '../utils'
import { defaultLogger } from '../logger'
import { dryRunExtrinsic, dryRunInherents } from './block-builder'

const logger = defaultLogger.child({ name: 'blockchain' })

export interface Options {
  api: Api
  buildBlockMode?: BuildBlockMode
  inherentProvider: InherentProvider
  db?: DataSource
  header: { number: number; hash: HexString }
  mockSignatureHost?: boolean
  allowUnresolvedImports?: boolean
  runtimeLogLevel?: number
  registeredTypes: RegisteredTypes
}

export class Blockchain {
  readonly uid: string = Math.random().toString(36).substring(2)
  readonly api: Api
  readonly db: DataSource | undefined
  readonly mockSignatureHost: boolean
  readonly allowUnresolvedImports: boolean
  readonly runtimeLogLevel: number
  readonly registeredTypes: RegisteredTypes

  readonly #txpool: TxPool
  readonly #inherentProvider: InherentProvider

  #head: Block
  readonly #blocksByNumber: Block[] = []
  readonly #blocksByHash: Record<string, Block> = {}
  readonly #loadingBlocks: Record<string, Promise<void>> = {}

  readonly headState: HeadState

  constructor({
    api,
    buildBlockMode,
    inherentProvider,
    db,
    header,
    mockSignatureHost = false,
    allowUnresolvedImports = false,
    runtimeLogLevel = 0,
    registeredTypes = {},
  }: Options) {
    this.api = api
    this.db = db
    this.mockSignatureHost = mockSignatureHost
    this.allowUnresolvedImports = allowUnresolvedImports
    this.runtimeLogLevel = runtimeLogLevel
    this.registeredTypes = registeredTypes

    this.#head = new Block(this, header.number, header.hash)
    this.#registerBlock(this.#head)

    this.#txpool = new TxPool(this, inherentProvider, buildBlockMode)
    this.#inherentProvider = inherentProvider

    this.headState = new HeadState(this.#head)
  }

  #registerBlock(block: Block) {
    this.#blocksByNumber[block.number] = block
    this.#blocksByHash[block.hash] = block
  }

  get head(): Block {
    return this.#head
  }

  get txPool() {
    return this.#txpool
  }

  async getBlockAt(number?: number): Promise<Block | undefined> {
    if (number === undefined) {
      return this.head
    }
    if (number > this.#head.number) {
      return undefined
    }
    if (!this.#blocksByNumber[number]) {
      const hash = await this.api.getBlockHash(number)
      const block = new Block(this, number, hash)
      this.#registerBlock(block)
    }
    return this.#blocksByNumber[number]
  }

  async getBlock(hash?: HexString): Promise<Block | undefined> {
    await this.api.isReady
    if (hash == null) {
      hash = this.head.hash
    }
    if (!this.#blocksByHash[hash]) {
      const loadingBlock = this.#loadingBlocks[hash]
      if (loadingBlock) {
        await loadingBlock
      } else {
        const loadingBlock = (async () => {
          try {
            const header = await this.api.getHeader(hash)
            const block = new Block(this, Number(header.number), hash)
            this.#registerBlock(block)
          } catch (e) {
            logger.debug(`getBlock(${hash}) failed: ${e}`)
          }
        })()
        this.#loadingBlocks[hash] = loadingBlock
        await loadingBlock
        delete this.#loadingBlocks[hash]
      }
    }
    return this.#blocksByHash[hash]
  }

  unregisterBlock(block: Block): void {
    if (block.hash === this.head.hash) {
      throw new Error('Cannot unregister head block')
    }
    if (this.#blocksByNumber[block.number]?.hash === block.hash) {
      delete this.#blocksByNumber[block.number]
    }
    delete this.#blocksByHash[block.hash]
  }

  async setHead(block: Block): Promise<void> {
    logger.debug(
      {
        number: block.number,
        hash: block.hash,
      },
      'setHead'
    )
    this.#head = block
    this.#registerBlock(block)
    await this.headState.setHead(block)
  }

  async submitExtrinsic(extrinsic: HexString): Promise<HexString> {
    const source = '0x02' // External
    const args = u8aToHex(u8aConcat(source, extrinsic, this.head.hash))
    const res = await this.head.call('TaggedTransactionQueue_validate_transaction', [args])
    const registry = await this.head.registry
    const validity: TransactionValidity = registry.createType('TransactionValidity', res.result)
    if (validity.isOk) {
      await this.#txpool.submitExtrinsic(extrinsic)
      return blake2AsHex(extrinsic, 256)
    }
    throw validity.asErr
  }

  submitUpwardMessages(id: number, ump: HexString[]) {
    this.#txpool.submitUpwardMessages(id, ump)

    logger.debug({ id, ump }, 'submitUpwardMessages')
  }

  submitDownwardMessages(dmp: DownwardMessage[]) {
    this.#txpool.submitDownwardMessages(dmp)

    logger.debug({ dmp }, 'submitDownwardMessages')
  }

  submitHorizontalMessages(id: number, hrmp: HorizontalMessage[]) {
    this.#txpool.submitHorizontalMessages(id, hrmp)

    logger.debug({ id, hrmp }, 'submitHorizontalMessages')
  }

  async newBlock(params?: Partial<BuildBlockParams>): Promise<Block> {
    await this.#txpool.buildBlock(params)
    return this.#head
  }

  async newBlockWithParams(params: BuildBlockParams): Promise<Block> {
    await this.#txpool.buildBlockWithParams(params)
    return this.#head
  }

  async upcomingBlocks() {
    return this.#txpool.upcomingBlocks()
  }

  async dryRunExtrinsic(
    extrinsic: HexString | { call: HexString; address: string },
    at?: HexString
  ): Promise<{ outcome: ApplyExtrinsicResult; storageDiff: [HexString, HexString | null][] }> {
    await this.api.isReady
    const head = at ? await this.getBlock(at) : this.head
    if (!head) {
      throw new Error(`Cannot find block ${at}`)
    }
    const registry = await head.registry
    const inherents = await this.#inherentProvider.createInherents(head, {
      transactions: [],
      downwardMessages: [],
      upwardMessages: [],
      horizontalMessages: {},
    })
    const { result, storageDiff } = await dryRunExtrinsic(head, inherents, extrinsic)
    const outcome = registry.createType<ApplyExtrinsicResult>('ApplyExtrinsicResult', result)
    return { outcome, storageDiff }
  }

  async dryRunHrmp(
    hrmp: Record<number, HorizontalMessage[]>,
    at?: HexString
  ): Promise<[HexString, HexString | null][]> {
    await this.api.isReady
    const head = at ? await this.getBlock(at) : this.head
    if (!head) {
      throw new Error(`Cannot find block ${at}`)
    }
    const inherents = await this.#inherentProvider.createInherents(head, {
      transactions: [],
      downwardMessages: [],
      upwardMessages: [],
      horizontalMessages: hrmp,
    })
    return dryRunInherents(head, inherents)
  }
  async dryRunDmp(dmp: DownwardMessage[], at?: HexString): Promise<[HexString, HexString | null][]> {
    await this.api.isReady
    const head = at ? await this.getBlock(at) : this.head
    if (!head) {
      throw new Error(`Cannot find block ${at}`)
    }
    const inherents = await this.#inherentProvider.createInherents(head, {
      transactions: [],
      downwardMessages: dmp,
      upwardMessages: [],
      horizontalMessages: {},
    })
    return dryRunInherents(head, inherents)
  }
  async dryRunUmp(ump: Record<number, HexString[]>, at?: HexString): Promise<[HexString, HexString | null][]> {
    await this.api.isReady
    const head = at ? await this.getBlock(at) : this.head
    if (!head) {
      throw new Error(`Cannot find block ${at}`)
    }
    const meta = await head.meta

    const needsDispatch = meta.registry.createType('Vec<u32>', Object.keys(ump))

    const stroageValues: [string, StorageValue | null][] = [
      [compactHex(meta.query.ump.needsDispatch()), needsDispatch.toHex()],
    ]

    for (const [paraId, messages] of Object.entries(ump)) {
      const upwardMessages = meta.registry.createType('Vec<Bytes>', messages)
      if (upwardMessages.length === 0) throw new Error('No upward meesage')

      const queueSize = meta.registry.createType('(u32, u32)', [
        upwardMessages.length,
        upwardMessages.map((x) => x.byteLength).reduce((s, i) => s + i, 0),
      ])

      stroageValues.push([compactHex(meta.query.ump.relayDispatchQueues(paraId)), upwardMessages.toHex()])
      stroageValues.push([compactHex(meta.query.ump.relayDispatchQueueSize(paraId)), queueSize.toHex()])
    }

    head.pushStorageLayer().setAll(stroageValues)
    const inherents = await this.#inherentProvider.createInherents(head, {
      transactions: [],
      downwardMessages: [],
      upwardMessages: [],
      horizontalMessages: {},
    })
    return dryRunInherents(head, inherents)
  }

  async getInherents(): Promise<HexString[]> {
    await this.api.isReady
    const inherents = await this.#inherentProvider.createInherents(this.head, {
      transactions: [],
      downwardMessages: [],
      upwardMessages: [],
      horizontalMessages: {},
    })
    return inherents
  }
}
