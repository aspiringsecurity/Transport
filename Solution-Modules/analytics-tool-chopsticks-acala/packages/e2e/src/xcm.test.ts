import { beforeEach, describe, it } from 'vitest'

import { DownwardMessage, HorizontalMessage } from '@acala-network/chopsticks/blockchain/txpool'
import { connectDownward } from '@acala-network/chopsticks/xcm/downward'
import { connectUpward } from '@acala-network/chopsticks/xcm/upward'
import { matchSystemEvents, testingPairs } from '@acala-network/chopsticks-testing'
import { setStorage } from '@acala-network/chopsticks/utils/set-storage'

import { matchSnapshot } from './helper'
import networks, { Network } from './networks'

const downwardMessages: DownwardMessage[] = [
  {
    sentAt: 1,
    msg: '0x0210010400010000078155a74e390a1300010000078155a74e39010300286bee0d01000400010100c0cbffafddbe39f71f0190c2369adfc59eaa4c81a308ebcad88cdd9c400ba57c',
  },
]

const horizontalMessages: Record<number, HorizontalMessage[]> = {
  2004: [
    {
      data: '0x000210000400000106080001000fc2ddd331d55e200a1300000106080001000fc2ddd331d55e20010700f2052a010d01000400010100ba686c8fa59178c699a698ea4d8e2c595394c2594bce4b6c2ca3a9bf3018e25d',
      sentAt: 13509121,
    },
  ],
}

describe('XCM', async () => {
  let acala: Network
  let polkadot: Network

  beforeEach(async () => {
    acala = await networks.acala()
    polkadot = await networks.polkadot()

    return async () => {
      await acala.teardown()
      await polkadot.teardown()
    }
  })

  it('Acala handles downward messages', async () => {
    await acala.chain.newBlock({ downwardMessages })
    await matchSystemEvents(acala)
  })

  it('Acala handles horizonal messages', async () => {
    await acala.chain.newBlock({ horizontalMessages })
    await matchSystemEvents(acala)
  })

  it('Polkadot send downward messages to Acala', async () => {
    await connectDownward(polkadot.chain, acala.chain)

    const { alice } = testingPairs()

    polkadot.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e10 } }]],
      },
    })

    await polkadot.api.tx.xcmPallet
      .reserveTransferAssets(
        { V0: { X1: { Parachain: 2000 } } },
        {
          V0: {
            X1: {
              AccountId32: {
                network: 'Any',
                id: alice.addressRaw,
              },
            },
          },
        },
        {
          V0: [
            {
              ConcreteFungible: { id: 'Null', amount: 100e10 },
            },
          ],
        },
        0
      )
      .signAndSend(alice)

    await polkadot.chain.newBlock()
    await matchSystemEvents(polkadot)

    await acala.chain.newBlock()
    await matchSystemEvents(acala)
  })

  it('Acala send upward messages to Polkadot', async () => {
    await connectUpward(acala.chain, polkadot.chain)

    const { alice } = testingPairs()

    await setStorage(acala.chain, {
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e10 } }]],
      },
      Tokens: {
        Accounts: [[[alice.address, { token: 'DOT' }], { free: 1000e10 }]],
      },
    })

    await matchSnapshot(polkadot.api.query.system.account(alice.address))
    await matchSnapshot(acala.api.query.system.account(alice.address))
    await matchSnapshot(acala.api.query.tokens.accounts(alice.address, { token: 'DOT' }))

    await acala.api.tx.xTokens
      .transfer(
        {
          Token: 'DOT',
        },
        10e10,
        {
          V1: {
            parents: 1,
            interior: {
              X1: {
                AccountId32: {
                  network: 'Any',
                  id: alice.addressRaw,
                },
              },
            },
          },
        },
        {
          Unlimited: null,
        }
      )
      .signAndSend(alice)

    await acala.chain.newBlock()
    await matchSystemEvents(acala)
    await matchSnapshot(acala.api.query.tokens.accounts(alice.address, { token: 'DOT' }))

    await polkadot.chain.newBlock()

    await matchSnapshot(polkadot.api.query.system.account(alice.address))
    await matchSystemEvents(polkadot)
  })
})
