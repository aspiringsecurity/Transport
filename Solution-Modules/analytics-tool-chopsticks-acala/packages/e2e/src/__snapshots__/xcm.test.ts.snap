// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XCM > Acala handles downward messages > system events 1`] = `
[
  {
    "data": {
      "amount": "(rounded 45000000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 2300000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1200000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 6800000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 260000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "count": 1,
    },
    "method": "DownwardMessagesReceived",
    "section": "parachainSystem",
  },
  {
    "data": {
      "amount": "(rounded 250000000000)",
      "currencyId": {
        "Token": "DOT",
      },
      "who": "25EPNyxpdqMUeVj9WndwUYVKkfFkMLroo2fg9V1MA79PJ6iE",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1700000)",
      "currencyId": {
        "Token": "DOT",
      },
      "who": "23M5ttkmR6KcoTAAE6gcmibnKFtVaTP5yxnY8HF1BmrJ2A1i",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "messageId": "(hash)",
      "outcome": {
        "Complete": 800000000,
      },
    },
    "method": "ExecutedDownward",
    "section": "dmpQueue",
  },
  {
    "data": {
      "dmqHead": "(hash)",
      "weightUsed": {
        "proofSize": 0,
        "refTime": 800000000,
      },
    },
    "method": "DownwardMessagesProcessed",
    "section": "parachainSystem",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "No",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 99000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
]
`;

exports[`XCM > Acala handles horizonal messages > system events 1`] = `
[
  {
    "data": {
      "amount": "(rounded 45000000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 2300000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1200000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 6800000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 260000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "amount": "(rounded 9100000000000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23UvQ3ZQXJ5LfTUSYkcRPkQX2FHgcKxGmqdxYJe9j5e3Lwsi",
    },
    "method": "Withdrawn",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 9100000000000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "2561WgnRM2Nbec2WseHxVdpeW19YYHh7RM5gRWQV7zwwwgXQ",
    },
    "method": "Endowed",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 9100000000000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "2561WgnRM2Nbec2WseHxVdpeW19YYHh7RM5gRWQV7zwwwgXQ",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 2100000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcoTAAE6gcmibnKFtVaTP5yxnY8HF1BmrJ2A1i",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "messageHash": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 800000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "No",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 99000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
]
`;

exports[`XCM > Acala send upward messages to Polkadot > system events 1`] = `
[
  {
    "data": {
      "amount": "(rounded 45000000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 2300000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1200000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 6800000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 260000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "No",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 99000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "amount": "(rounded 2600000000)",
      "who": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "Withdraw",
    "section": "balances",
  },
  {
    "data": {
      "amount": 100000000000,
      "currencyId": {
        "Token": "DOT",
      },
      "who": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "Withdrawn",
    "section": "tokens",
  },
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 100000000000,
          },
          "id": {
            "Concrete": {
              "interior": "Here",
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X1": {
            "AccountId32": {
              "id": "(hash)",
              "network": "Any",
            },
          },
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 100000000000,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
      "sender": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
  {
    "data": {
      "amount": "(rounded 2600000000)",
      "who": "23M5ttkmR6KcoTAAE6gcmibnKFtVaTP5yxnY8HF1BmrJ2A1i",
    },
    "method": "Deposit",
    "section": "balances",
  },
  {
    "data": {
      "value": "(rounded 2600000000)",
    },
    "method": "Deposit",
    "section": "treasury",
  },
  {
    "data": {
      "actualFee": "(rounded 2600000000)",
      "actualSurplus": 0,
      "actualTip": 0,
      "who": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "TransactionFeePaid",
    "section": "transactionPayment",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 600000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
]
`;

exports[`XCM > Acala send upward messages to Polkadot > system events 2`] = `
[
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 230000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "amount": 100000000000,
      "who": "13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy",
    },
    "method": "Withdraw",
    "section": "balances",
  },
  {
    "data": {
      "amount": "(rounded 100000000000)",
      "who": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
    },
    "method": "Deposit",
    "section": "balances",
  },
  {
    "data": {
      "account": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
    },
    "method": "NewAccount",
    "section": "system",
  },
  {
    "data": {
      "account": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
      "freeBalance": "(rounded 100000000000)",
    },
    "method": "Endowed",
    "section": "balances",
  },
  {
    "data": {
      "amount": "(rounded 410000000)",
      "who": "12ZNJzjPBZUh8VV5cuJFkbbwMttFNkH39EhoeYcgGHsJd4MG",
    },
    "method": "Deposit",
    "section": "balances",
  },
  {
    "data": [
      "(hash)",
      {
        "Complete": 4000000000,
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 98000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
]
`;

exports[`XCM > Acala send upward messages to Polkadot 1`] = `
{
  "consumers": "0",
  "data": {
    "feeFrozen": "0",
    "free": "0",
    "miscFrozen": "0",
    "reserved": "0",
  },
  "nonce": "0",
  "providers": "0",
  "sufficients": "0",
}
`;

exports[`XCM > Acala send upward messages to Polkadot 2`] = `
{
  "consumers": "0",
  "data": {
    "feeFrozen": "0",
    "free": "10,000,000,000,000",
    "miscFrozen": "0",
    "reserved": "0",
  },
  "nonce": "0",
  "providers": "0",
  "sufficients": "0",
}
`;

exports[`XCM > Acala send upward messages to Polkadot 3`] = `
{
  "free": "10,000,000,000,000",
  "frozen": "0",
  "reserved": "0",
}
`;

exports[`XCM > Acala send upward messages to Polkadot 4`] = `
{
  "free": "9,900,000,000,000",
  "frozen": "0",
  "reserved": "0",
}
`;

exports[`XCM > Acala send upward messages to Polkadot 5`] = `
{
  "consumers": "0",
  "data": {
    "feeFrozen": "0",
    "free": "99,591,353,032",
    "miscFrozen": "0",
    "reserved": "0",
  },
  "nonce": "0",
  "providers": "1",
  "sufficients": "0",
}
`;

exports[`XCM > Polkadot send downward messages to Acala > system events 1`] = `
[
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 230000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 98000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "amount": "(rounded 180000000)",
      "who": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
    },
    "method": "Withdraw",
    "section": "balances",
  },
  {
    "data": {
      "amount": 1000000000000,
      "from": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
      "to": "13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy",
    },
    "method": "Transfer",
    "section": "balances",
  },
  {
    "data": [
      {
        "Complete": 1000000000,
      },
    ],
    "method": "Attempted",
    "section": "xcmPallet",
  },
  {
    "data": {
      "amount": "(rounded 150000000)",
      "who": "13UVJyLnbVp9RBZYFwFGyDvVd1y27Tt8tkntv6Q7JVPhFsTB",
    },
    "method": "Deposit",
    "section": "balances",
  },
  {
    "data": {
      "value": "(rounded 150000000)",
    },
    "method": "Deposit",
    "section": "treasury",
  },
  {
    "data": {
      "amount": "(rounded 37000000)",
      "who": "12ZNJzjPBZUh8VV5cuJFkbbwMttFNkH39EhoeYcgGHsJd4MG",
    },
    "method": "Deposit",
    "section": "balances",
  },
  {
    "data": {
      "actualFee": "(rounded 180000000)",
      "tip": 0,
      "who": "146SvjUZXoMaemdeiecyxgALeYMm8ZWh1yrGo8RtpoPfe7WL",
    },
    "method": "TransactionFeePaid",
    "section": "transactionPayment",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 1200000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
]
`;

exports[`XCM > Polkadot send downward messages to Acala > system events 2`] = `
[
  {
    "data": {
      "amount": "(rounded 45000000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 2300000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1200000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 6800000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 260000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "count": 1,
    },
    "method": "DownwardMessagesReceived",
    "section": "parachainSystem",
  },
  {
    "data": {
      "account": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "NewAccount",
    "section": "system",
  },
  {
    "data": {
      "amount": "(rounded 1000000000000)",
      "currencyId": {
        "Token": "DOT",
      },
      "who": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "Endowed",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1000000000000)",
      "currencyId": {
        "Token": "DOT",
      },
      "who": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1700000)",
      "currencyId": {
        "Token": "DOT",
      },
      "who": "23M5ttkmR6KcoTAAE6gcmibnKFtVaTP5yxnY8HF1BmrJ2A1i",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "messageId": "(hash)",
      "outcome": {
        "Complete": 800000000,
      },
    },
    "method": "ExecutedDownward",
    "section": "dmpQueue",
  },
  {
    "data": {
      "dmqHead": "(hash)",
      "weightUsed": {
        "proofSize": 0,
        "refTime": 800000000,
      },
    },
    "method": "DownwardMessagesProcessed",
    "section": "parachainSystem",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "No",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 99000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
]
`;
